### PR TITLE
Add huggingface based models exporter

### PR DESCRIPTION
**Pull request by ktagowski**
TODO:

- [x] Add option to export model in `ONNX` format
- [x] Fix mypy
- [x] Write tests
~~- [ ] Wrap huggingface model upload (separate PR)~~

Example usage (POC)
```python

from embeddings.config.lightning_config import LightningBasicConfig
from embeddings.pipeline.lightning_classification import LightningClassificationPipeline
from embeddings.utils.hf_model_exporter import HuggingFaceModelExporter

pipeline = LightningClassificationPipeline(
    embedding_name_or_path="hf-internal-testing/tiny-albert",
    devices="auto",
    accelerator="cpu",
    output_path=".",
    dataset_name_or_path="clarin-pl/polemo2-official",
    input_column_name=["text"],
    target_column_name="target",
    config=LightningBasicConfig(max_epochs=1)
)
result = pipeline.run()
HuggingFaceModelExporter("hf_model_export_test").export_model_from_pipeline(pipeline)
```

### CODE REVIEW

File: embeddings/utils/model_exporter.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    def _check_task(self) -> None:
+        model = getattr(self._task_to_export.model, "model", None)
+        if not model:
+            raise LightningTask.MODEL_UNDEFINED_EXCEPTION
+
+    @staticmethod
+    def _map_target_names(model: AutoModel, target_names: List[str]) -> AutoModel:
+        if target_names:
+            id2label = {k: v for k, v in zip(model.config.id2label.keys(), target_names)}
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/utils/model_exporter.py:56]: Assertion added

File: embeddings/utils/model_exporter.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        self._tokenizer_to_export = tokenizer
+
+        label_names = getattr(task, "target_names", None)
+        if label_names:
+            self._model_to_export = BaseModelExporter._map_target_names(
+                self._model_to_export, target_names=label_names
+            )
+
+        hparams = getattr(self._task_to_export.model, "hparams", None)
+        if self.add_hparams_configuration and not hparams:
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply embeddings/utils/model_exporter.py:98]: I added dedicated error for that and in same way it is raised where tokenizer is missing

File: pyproject.toml [hunk 3]
```diff
@@ -129,7 +130,9 @@ module = [
     "dataset.arrow_dataset",
     "seqeval.*",
     "transformers",
-    "evaluate"
+    "transformers.*",
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply pyproject.toml]: Fixed

File: tests/test_model_exporter.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+            getattr(pretrained_metrics, metric_key),
+            getattr(results, metric_key),
+            decimal=pytest.decimal,
+        )
+
+
+def test_hf_model_exporter_from_task(
+    embedding_name: str,
+    lightning_text_classification_pipeline: Tuple[
+        LightningClassificationPipeline, TextClassificationEvaluationResults
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply tests/test_model_exporter.py]: I defined type for that

File: tests/test_model_exporter.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+            "attention_mask": batch["attention_mask"].numpy(),
+        }
+        outputs = session.run(output_names=["logits"], input_feed=np_batch)[0]
+        preds += torch.argmax(torch.Tensor(outputs), dim=1)
+        y_true += batch["labels"]
+
+    y_pred = torch.IntTensor(preds).numpy()
+    y_true = torch.IntTensor(y_true).numpy()
+    predictions = Predictions(y_pred=y_pred, y_true=y_true)
+    results = TextClassificationEvaluator().evaluate(predictions)
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply tests/test_model_exporter.py]: Fixed

File: tests/test_model_exporter.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+    ],
+    config: LightningAdvancedConfig,
+    tmp_path_module: Path,
+    dataset_kwargs: Dict[str, Any],
+    comparison_metrics_keys: List[str],
+    hf_datamodule: TextClassificationDataModule,
+):
+    path = tmp_path_module / "onnx_model_task"
+    exporter = ONNXModelExporter(path=path)
+    pretrained_pipeline = lightning_text_classification_pipeline[0]
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply tests/test_model_exporter.py]: I was so focused on providing appriopate model inference code that I put not so much effort in writing clean code. Fixed

File: tests/test_model_exporter.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    pretrained_metrics = lightning_text_classification_pipeline[1]
+    exporter.export_model_from_task(pretrained_pipeline.model.task)
+    results = evaluate_onnx_model(model_path=path, datamodule=hf_datamodule)
+
+    for metric_key in comparison_metrics_keys:
```
[the remark this answers is not in the document]
Comment (ktagowski) [reply tests/test_model_exporter.py]: Yes, but that require additional package like: pytest-check or pytest-assume. It's nice idea but I would handle it in seperate PR as I would improve tests in also other places

**Review by laugustyniak — DISMISSED** [review]

**Review by laugustyniak — APPROVED** [review]

**Pull request merged**